jira/jira.go

```diff
@@ -1,41 +1,110 @@
 package jira
 
 import (
+	"errors"
+	gojira "github.com/andygrunwald/go-jira"
 	"github.com/go-chat-bot/bot"
+	"log"
 	"os"
 	"regexp"
-	"strings"
 )
 
 const (
-	pattern = "(^|\\s)+[A-z]{3}-[0-9]+\\b"
-	env     = "JIRA_ISSUES_URL"
+	pattern    = ".*?([A-Z]+)-([0-9]+)\\b"
+	env        = "JIRA_ISSUES_URL"
+	userEnv    = "JIRA_USER"
+	passEnv    = "JIRA_PASS"
+	baseURLEnv = "JIRA_BASE_URL"
 )
 
 var (
-	url string
-	re  = regexp.MustCompile(pattern)
+	url      string
+	baseURL  string
+	jiraUser string
+	jiraPass string
+	projects map[string]gojira.Project
+	client   *gojira.Client
+	re       = regexp.MustCompile(pattern)
 )
 
-func getIssue(text string) string {
-	if re.MatchString(text) {
-		return re.FindString(text)
+func getProjects() (map[string]gojira.Project, error) {
+	req, err := client.NewRequest("GET", "rest/api/2/project", nil)
+	if err != nil {
+		return projects, errors.New("Error creating request object: " + err.Error())
 	}
-	return ""
+
+	projectObjects := new([]gojira.Project)
+	projects = make(map[string]gojira.Project)
+	_, err = client.Do(req, projectObjects)
+	if err != nil {
+		return projects, errors.New("Failed getting JIRA projects: " + err.Error())
+	}
+	for _, project := range *projectObjects {
+		projects[project.Key] = project
+	}
+	return projects, nil
 }
 
-func jira(cmd *bot.PassiveCmd) (string, error) {
-	issue := getIssue(cmd.Raw)
-	if issue != "" {
-		return url + strings.ToUpper(strings.TrimSpace(issue)), nil
+func getIssues(text string) [][2]string {
+	matches := re.FindAllStringSubmatch(text, -1)
+	var data [][2]string
+	for _, match := range matches {
+		// match[1] == project key
+		// match[2] == issue number
+		data = append(data, [2]string{match[1], match[2]})
 	}
+	return data
+}
 
-	return "", nil
+func jira(cmd *bot.PassiveCmd) (bot.CmdResultV3, error) {
+	result := bot.CmdResultV3{
+		Message: make(chan string),
+		Done:    make(chan bool, 1)}
+	result.Channel = cmd.Channel
+	issues := getIssues(cmd.Raw)
+	if issues != nil {
+		go func() {
+			for _, issue := range issues {
+				key, num := issue[0], issue[1]
+				_, found := projects[key]
+				if found {
+					result.Message <- url + key + "-" + num
+				}
+			}
+			result.Done <- true
+		}()
+	} else {
+		result.Done <- true
+	}
+
+	return result, nil
 }
 
 func init() {
+	var err error
 	url = os.Getenv(env)
-	bot.RegisterPassiveCommand(
+	jiraUser = os.Getenv(userEnv)
+	jiraPass = os.Getenv(passEnv)
+	baseURL = os.Getenv(baseURLEnv)
+
+	tp := gojira.BasicAuthTransport{
+		Username: jiraUser,
+		Password: jiraPass,
+	}
+
+	client, err = gojira.NewClient(tp.Client(), baseURL)
+	if err != nil {
+		log.Printf("Error initializing JIRA client: %s ", err)
+		return
+	}
+
+	_, err = getProjects()
+	if err != nil {
+		log.Printf("Error querying JIRA for projects: " + err.Error())
```

Comment:
You can use the format insteat of `+`

```Go
log.Printf("Error querying JIRA for projects: %v\n", err))
```

Also notice that Printf doesn't add new lines, so always end it with a `\n`